tests/test_preorder_walk.py

```diff
@@ -0,0 +1,119 @@
+                                "test.finish"() : () -> ()
+                },
+                {
+                        ^0:
+                                "test.variadic_no_terminator_op"() {"n_op"=3} : () -> ()
+                                "test.finish"() : () -> ()
+                }) : () -> ()
+                "test.finish"() : () -> ()
+        },
+        {
+        ^0:
+                "test.variadic_no_terminator_op"() {"n_op"=4} : () -> ()
+                "test.single_region_op"() ({
+                        ^0:
+                                "test.variadic_no_terminator_op"() {"n_op"=5} : () -> ()
+                                "test.two_region_op"() ({
+                                        "test.variadic_no_terminator_op"() {"n_op"=6} : () -> ()
+                                        "test.finish"() : () -> ()
+                                },
+                                {
+                                        "test.variadic_no_terminator_op"() {"n_op"=7} : () -> ()
+                                        "test.finish"() : () -> ()
+
+                                }) : () -> ()
+                                "test.finish"() : () -> ()
+                }) : () -> ()
+                "test.finish"() : () -> ()
+        }) : () -> ()
+        "test.finish"() : () -> ()
+}) : () -> ()
+"""
+
+test_prog = """
+"test.op"() ({
+^0:
+       "test.op"() ({
+       ^0:
+               "test.op"() {"n_op"=1} : () -> ()
+               "test.op"() ({
+                       ^0:
+                               "test.op"() {"n_op"=2} : () -> ()
+                               "test.termop"() : () -> ()
+               },
+               {
+                       ^0:
+                               "test.op"() {"n_op"=3} : () -> ()
+                               "test.termop"() : () -> ()
+               }) : () -> ()
+               "test.termop"() : () -> ()
+       },
+       {
+       ^0:
+               "test.op"() {"n_op"=4} : () -> ()
+               "test.op"() ({
+                       ^0:
+                               "test.op"() {"n_op"=5} : () -> ()
+                               "test.op"() ({
+                                       "test.op"() {"n_op"=6} : () -> ()
+                                       "test.termop"() : () -> ()
+                               },
+                               {
+                                       "test.op"() {"n_op"=7} : () -> ()
+                                       "test.termop"() : () -> ()
+                               }) : () -> ()
+                               "test.termop"() : () -> ()
+               }) : () -> ()
+               "test.termop"() : () -> ()
+       }) : () -> ()
+       "test.termop"() : () -> ()
+}) : () -> ()
+"""
+
+
+def test_preorder_walk():
+    ctx = MLContext()
+    ctx.load_dialect(Test)
+
+    parser = Parser(ctx, test_prog)
+    op = parser.parse_op()
```

Comment:
sorry to be annoying, but do you think you could change this from parsing the text to constructing it in Python? Something like [this](https://github.com/xdslproject/xdsl/blob/main/tests/transforms/test_loop_invariant_code_motion.py) would make the test a little easier to read and understand, IMO. then we could just `assert tuple(op.walk_blocks()) == (...)`